lib/ClangImporter/ClangImporter.cpp

```diff
@@ -2962,12 +2962,36 @@ void ClangImporter::loadExtensions(NominalTypeDecl *nominal,
   // For an Objective-C class, import all of the visible categories.
   if (auto objcClass = dyn_cast_or_null<clang::ObjCInterfaceDecl>(
                          effectiveClangContext.getAsDeclContext())) {
+    SmallVector<clang::NamedDecl *, 4> DelayedCategories;
+
     // Simply importing the categories adds them to the list of extensions.
     for (auto I = objcClass->visible_categories_begin(),
            E = objcClass->visible_categories_end();
          I != E; ++I) {
+      // Delay installing categories that don't have an owning module.
+      if (!I->getOwningModule()) {
```

Comment:
Are all categories defined outside of the module "by us"? So this would catch only extensions exported to Objective-C or do you expect that it will affect other categories?

Reply:
Categories with no Clang Module are synthesized by us - usually in textual headers like bridging headers.  The worst I expect to happen is somebody has such a setup and somehow managed to not get the re-entrant behavior and so saw the decls in the textual header stomp the ones in the module.

Reply:
BTW, there should also be a `hasOwningModule` that may be faster.

Reply:
Ah, true.  I'll get that in a follow-up.